Generate JSON files to be consumed by Vue.js UI

dan-niles opened the pull request:
This PR adds the functionality for generating JSON files to be consumed by Vue.js UI as mentioned in #212. The JSON files being generated are:
- `channel.json` : Contains information about the channel
- `playlists.json` : Contains a list of all playlists in the channel
- Seperate JSON files under `playlists` folder for each playlist : Contains info about the playlist and the videos it contains. (This includes the "Uploads" of the channel as well)
- Seperate JSON files under `videos` folder for each video : Contains info about the specific video

The JSON files are added "on the fly" to the ZIM as mentioned in #209. However the video files and branding files are still being copied from the build_dir.

Closes #212
Closes #220

benoit74 commented on scraper/src/youtube2zim/scraper.py:
```diff
@@ -1108,3 +1120,240 @@ def to_data_js(video):
+
+        def generate_video_object(video):
+            video_id = video["contentDetails"]["videoId"]
+            title = video["snippet"]["title"]
```
title, description and publication_date are used only once, could be used directly in object, isn't it?

benoit74 commented on scraper/src/youtube2zim/scraper.py:
```diff
@@ -1108,3 +1120,240 @@ def to_data_js(video):
+            zim Creator
+        fpath:
+            path to the file/folder to add to the ZIM"""
+        if fpath.is_dir():
```
Maybe add something like `logger.debug("Adding files to ZIM")` so that next lines are clearer

benoit74 commented on scraper/src/youtube2zim/scraper.py:
```diff
@@ -1108,3 +1120,240 @@ def to_data_js(video):
+        self.zim_file.add_item_for(
+            path="playlists.json",
+            title="Playlists",
+            content=json.dumps(playlist_list, default=pydantic_encoder, indent=2),
```
Rather than using this different method (everywhere else you use `model_dump_json`) I would rather add a `Playlists` class to the schemas, with for now only one property named `items` which is the list of playlist.

This avoid the specifity, allows for greater possibility of extension (should we need more than simply the list of playlists in this file) and avoids security issues (I'm not a specialist, but it looks like it is not recommended to have a list as JSON file content).

dan-niles replied:
Changed in [25b7546](https://github.com/openzim/youtube/pull/219/commits/25b7546cc4a3657c357a8ce3e8acb2fb28c90439)

benoit74 commented on scraper/src/youtube2zim/scraper.py:
```diff
@@ -1108,3 +1120,240 @@ def to_data_js(video):
+            videos = sorted(videos, key=lambda v: v["snippet"]["position"])
+            return videos
+
+        def generate_video_object(video):
```
Add type hint (at least for return value, input is probably not really typed yet)

benoit74 commented on scraper/src/youtube2zim/scraper.py:
```diff
@@ -1108,3 +1120,240 @@ def to_data_js(video):
+                duration=0,
+            )
+
+        def generate_video_preview_object(video):
```
Type hint for returned value again

dan-niles replied:
Added in [e62b0bb](https://github.com/openzim/youtube/pull/219/commits/e62b0bb2b30b069f1377f4ce296ee69f2852f833)

benoit74 commented on scraper/src/youtube2zim/scraper.py:
```diff
@@ -1108,3 +1120,240 @@ def to_data_js(video):
+                duration=0,
+            )
+
+        def get_video_slug(video):
```
Type hint for returned value again

dan-niles replied:
Added in [e62b0bb](https://github.com/openzim/youtube/pull/219/commits/e62b0bb2b30b069f1377f4ce296ee69f2852f833)

benoit74 commented on scraper/src/youtube2zim/scraper.py:
```diff
@@ -1108,3 +1120,240 @@ def to_data_js(video):
+            video_id = video["contentDetails"]["videoId"]
+            return f"{get_slug(title)}-{video_id[:4]}"
+
+        def generate_playlist_object(playlist):
```
Type hint for returned value again

dan-niles replied:
Added in [e62b0bb](https://github.com/openzim/youtube/pull/219/commits/e62b0bb2b30b069f1377f4ce296ee69f2852f833)

benoit74 commented on scraper/src/youtube2zim/scraper.py:
```diff
@@ -1108,3 +1120,240 @@ def to_data_js(video):
+                id=playlist.playlist_id,
+                title=playlist.title,
+                description=playlist.description,
+                videos=videos_list,
```
I would rewrite it with something like (not tested) `videos = [generate_video_preview_object(video) for video in get_videos_list(playlist)]` (this one-liner get is both easier to read and more concise)

dan-niles replied:
Refactored in [8ff4342](https://github.com/openzim/youtube/pull/219/commits/8ff43429631f402e4e51d66853f9aa2f88ef8d7f)

benoit74 reviewed (changes requested):
Thank you.

Main changes required are around the usage of type hints. I know mostly no type hints are used so far in current codebase, but we would like to have as many as possible. So please try at least to add as many as reasonably possible in the code you add or modify.

Second concern is that I don't know how to test this, do we have at least a test channel or set of playlists where we could check how this code behaves? If not, we should probably build one so that we avoid reinventing the test scenario every time we test this. It should be small enough so that we can scrape it quite quickly, but big enough to exercise most scenarii of the scraper.

Finally, I'm a bit ashamed to continue to create code with absolutely no tests. I don't know if you have any suggestion regarding how we could automate testing this at least a bit. Maybe an issue for later, scraper has worked without tests for year, we can maybe continue like this for few more years ^^

dan-niles commented on CHANGELOG:
```diff
@@ -12,6 +12,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Bump `requests` package from 2.32.0 to 2.32.2
 - Initialize new Vue.js project in `zimui` subfolder
+- Update dependencies in pyproject.toml (pydantic, pyhumps, python-slugify)
+- Update scraper to generate JSON files for `zimui`
```
Added in [b92db90](https://github.com/openzim/youtube/pull/219/commits/b92db90c57df8e52fd672e8c77b3f4872d788ed9)

dan-niles commented:
@benoit74 I've included a zimdump of a ZIM file I generated using this new code here.  Can you check the JSON files and let me know if everything is ok?

[zimdump.zip](https://github.com/user-attachments/files/15566659/zimdump.zip)

(I removed the video files since there is a limit of 25MB for things I can upload here)

dan-niles commented:
Additionally, you can try running the scraper on the new testing channel created in #221 with the following command:

```
youtube2zim --api-key "<YOUR-API-KEY>" --type channel --id "UC8elThf5TGMpQfQc_VE917Q" --name "openZIM_testing" --output ../output
```

dan-niles commented:
I added a new `videoSlug` property to the `PlaylistPreview` schema, that points to the video to be redirected to when clicking on the preview. Also, added `joined_date` to channel.json.

benoit74 commented on scraper/src/youtube2zim/schemas.py:
```diff
@@ -0,0 +1,94 @@
+    publication_date: str
+    thumbnail_path: str | None = None
+    videos: list[VideoPreview]
+    count: int
```
Rename it to `videos_count` to remove any ambiguities

dan-niles replied:
Renamed in [1193985](https://github.com/openzim/youtube/pull/219/commits/119398503ebcc30b18d5accb33b46033d957ae90)

benoit74 commented on scraper/src/youtube2zim/schemas.py:
```diff
@@ -0,0 +1,94 @@
+    id: str
+    title: str
+    thumbnail_path: str | None = None
+    count: int
```
Rename it to `videos_count` to remove any ambiguities

dan-niles replied:
Renamed in [1193985](https://github.com/openzim/youtube/pull/219/commits/119398503ebcc30b18d5accb33b46033d957ae90)

benoit74 commented on scraper/src/youtube2zim/schemas.py:
```diff
@@ -0,0 +1,94 @@
+    title: str
+    thumbnail_path: str | None = None
+    count: int
+    video_slug: str
```
Should be renamed `main_video_slug` or `preview_video_slug` if I'm not mistaken

dan-niles replied:
Renamed in [1193985](https://github.com/openzim/youtube/pull/219/commits/119398503ebcc30b18d5accb33b46033d957ae90)

dan-niles replied:
Renamed in [1193985](https://github.com/openzim/youtube/pull/219/commits/119398503ebcc30b18d5accb33b46033d957ae90)

benoit74 reviewed (changes requested):
Some more things to fix (I'm still very pleased by the PR, I just dug a bit more and found more stuff ^^):
- see inline comments
- remove what we now know for sure is obsolete code and files: 
  - `make_html_files` function (and maybe others I've missed)
  - the whole `templates` folder and files
- translation of the UI was pretty minimalist, most probably partially broken, now it is completely broken by this PR (and previous one) ; I suggest to remove all stuff linked to the previous translation since it is just useless/missleading now (`locale` subfolder + installation of `locales-all` in Dockerfile) and open a new issue "Create structure for new Youtube UI translations" (I can open the issue with details if you want)

dan-niles commented on scraper/src/youtube2zim/scraper.py:
```diff
@@ -1108,3 +1120,238 @@ def to_data_js(video):
+        # clean videos left out in videos directory
+        remove_unused_videos(videos)
+
+    def add_files_to_zim(self, root: Path, zim_file: Creator, fpath: Path):
```
entrypoint.py already has arg called "[creator](https://github.com/openzim/youtube/blob/main/scraper/src/youtube2zim/entrypoint.py#L135)" that gets mapped to `self.creator`, that's why I named it `zim_file`. Shall I rename it to `zim_creator`?

benoit74 replied:
As discussed live, let's keep `zim_file`, it makes sense indeed

dan-niles replied:
Refactored `add_files_to_zim`  in [34aa44a](https://github.com/openzim/youtube/pull/219/commits/34aa44a02c95541173c0dd1399031ab301eb6855)

dan-niles commented:
Removed obsolete code and files:
- `make_html_files` and related imports in [a605e72](https://github.com/openzim/youtube/pull/219/commits/a605e7216855946c8f6953e3bc76684b0cb1410b)
- Unused templates (home.html, article.html) in [a605e72](https://github.com/openzim/youtube/pull/219/commits/a605e7216855946c8f6953e3bc76684b0cb1410b)
- `locale` folder in [725720d](https://github.com/openzim/youtube/pull/219/commits/725720dea8ad1d7dfde940a557185d7ab87374aa)

benoit74 commented on CHANGELOG:
```diff
@@ -14,6 +14,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Initialize new Vue.js project in `zimui` subfolder
 - Update dependencies in pyproject.toml (pydantic, pyhumps, python-slugify)
 - Update scraper to generate JSON files for `zimui` (#212)
+- Remove template files (home.html, article.html) and `make_html_files` method in scraper.py
```
Please be more specific with something like `Remove old UI files and methods: template files (home.html, article.html) and `make_html_files` method in scraper.py`

dan-niles replied:
Changed in [9981785](https://github.com/openzim/youtube/pull/219/commits/99817858463334030f4807a0ec5f975ac63189c8)

benoit74 commented on CHANGELOG:
```diff
@@ -14,6 +14,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Update dependencies in pyproject.toml (pydantic, pyhumps, python-slugify)
 - Update scraper to generate JSON files for `zimui` (#212)
+- Remove template files (home.html, article.html) and `make_html_files` method in scraper.py
+- Remove locale folder and files
```
Please be more specific with something like `Remove broken locale folder and files used for translation ; translation will be restored with #222`

dan-niles replied:
Changed in [0d7050f](https://github.com/openzim/youtube/pull/219/commits/0d7050f2ff8a5e1a5b8d17cebe5e784fe1a4cf42)

benoit74 reviewed (changes requested):
Two more small comments regarding the CHANGELOG.

Please take them into account and you are now good to go to rewrite the Git history of your branch to keep only a handful of meaningful commits, as discussed yesterday. I don't really mind about the exact number, but I don't want to keep stuff which is only linked to the back-and-forth exchanges we had on this PR and do not really make much sense for the history (long term goal of commit history is to understand why changes have been done, not how). Force push will be needed after the rewrite obviously. Do not hesitate to ask for help if what I expect is too blurry for you or if you face issues.

Ask for a new review once this has been done and I will merge this.

benoit74 reviewed (approved):
LGTM, thank you!
